# Capstone Project: Late Delivery Risk Scoring in Global Supply Chains
**Author/Team:** Aman Abuelayyan, Ameed Aburub, Dalah Hashlamoon, Ruaa Luay, and Sura Sammour.

**Date:** March, 11th 2026.

**Track:** Supervised Learning (Classification + Regression)

### **Project Objective**
To predict how late an order will be and combine that with its financial impact to produce a single, actionable continuous "Risk Score" for supply chain managers.

### **Table of Contents**
1. [Setup and Imports](#setup)
2. [Data Loading & Merging](#loading)
3. [Data Cleaning & Formatting](#cleaning)
4. [Exploratory Data Analysis (EDA)](#eda)
5. [Summary & Next Steps](#summary)

## 1. Setup and Imports <a id="setup"></a>
The necessary libraries and steps for the notbook are made first.

### 1.1. Import drive directory

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

### 1.2. Import & configure necessary libraries

In [ ]:
import warnings
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

warnings.filterwarnings('ignore')

# Consistent visual style
sns.set_theme(style='whitegrid', palette='muted', font_scale=1.1)
plt.rcParams['figure.dpi'] = 110
plt.rcParams['axes.spines.top']   = False
plt.rcParams['axes.spines.right'] = False

print('Libraries loaded.')

**Library Imports:** It imports core data science libraries including pandas and numpy for data manipulation, and matplotlib and seaborn for data visualization.

**Configuration:** It sets a default visualization style (whitegrid), defines a standard figure size for plots, and suppresses warnings to ensure a clean output in the notebook.

## 2. Data Loading & Understanding <a id="loading"></a>
The dataset is transformed into a single DataFrame.

### 2.1. Load the data

In [ ]:
# Load data CSV into a DataFrame.
df_raw = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/DataCoSupplyChainDataset.csv', encoding='latin1')
print(f'Loaded {len(df_raw):,} rows x {df_raw.shape[1]} columns')

df_raw.head(5)

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Cally,20755,Holloway,XXXXXXXXX,Consumer,PR,5365 Noble Nectar Island,725.0,2,Fitness,18.251453,-66.037056,Pacific Asia,Bekasi,Indonesia,20755,1/31/2018 22:56,77202,1360,13.110000,0.04,180517,327.75,0.29,1,327.75,314.640015,91.250000,Southeast Asia,Java Occidental,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,02-03-18 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Irene,19492,Luna,XXXXXXXXX,Consumer,PR,2679 Rustic Loop,725.0,2,Fitness,18.279451,-66.037064,Pacific Asia,Bikaner,India,19492,1/13/2018 12:27,75939,1360,16.389999,0.05,179254,327.75,-0.80,1,327.75,311.359985,-249.089996,South Asia,Rajastán,PENDING,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,EE. UU.,XXXXXXXXX,Gillian,19491,Maldonado,XXXXXXXXX,Consumer,CA,8510 Round Bear Gate,95125.0,2,Fitness,37.292233,-121.881279,Pacific Asia,Bikaner,India,19491,1/13/2018 12:06,75938,1360,18.030001,0.06,179253,327.75,-0.80,1,327.75,309.720001,-247.779999,South Asia,Rajastán,CLOSED,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,EE. UU.,XXXXXXXXX,Tana,19490,Tate,XXXXXXXXX,Home Office,CA,3200 Amber Bend,90027.0,2,Fitness,34.125946,-118.291016,Pacific Asia,Townsville,Australia,19490,1/13/2018 11:45,75937,1360,22.940001,0.07,179252,327.75,0.08,1,327.75,304.809998,22.860001,Oceania,Queensland,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Orli,19489,Hendricks,XXXXXXXXX,Corporate,PR,8671 Iron Anchor Corners,725.0,2,Fitness,18.253769,-66.037048,Pacific Asia,Townsville,Australia,19489,1/13/2018 11:24,75936,1360,29.500000,0.09,179251,327.75,0.45,1,327.75,298.250000,134.210007,Oceania,Queensland,PENDING_PAYMENT,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


**Data Loading:** It reads a CSV file named data.csv from the /content/ directory into a pandas DataFrame using latin1 encoding.

**Data Inspection:** It calls df.head(5) to display the first five rows of the dataset, allowing for a quick check of the column names and initial values.

### 2.2. Data shape

In [ ]:
df_raw.shape

(180519, 53)

**Shape Inspection:** It uses df.shape to return a tuple representing the dimensionality of the DataFrame. Based on the output, the dataset contains 180,519 rows and 53 columns.

### 2.3. Data information

In [ ]:
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  object 
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  object 
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  object 
 9   Customer City                  180519 non-null  object 
 10  Customer Country               180519 non-null  object 
 11  Customer Email                 180519 non-null  object 
 12  Customer Fname                

**Metadata Inspection:** The `df.info()` method provides a concise summary of the DataFrame, including the number of entries, column names, non-null counts, and data types (e.g., float64, int64, object). This is essential for identifying missing values and understanding the data structure.

### 2.4. Data description

In [ ]:
df_raw.describe()

,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Late_delivery_risk,Category Id,Customer Id,Customer Zipcode,Department Id,Latitude,Longitude,Order Customer Id,Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Price,Product Status
count,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180516.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,24840.000000,180519.000000,180519.000000,0.0,180519.000000,180519.0
mean,3.497654,2.931847,21.974989,183.107609,0.548291,31.851451,6691.379495,35921.126914,5.443460,29.719955,-84.915675,6691.379495,36221.894903,692.509764,20.664741,0.101668,90260.000000,141.232550,0.120647,2.127638,203.772096,183.107609,21.974989,55426.132327,692.509764,31.851451,NaN,141.232550,0.0
std,1.623722,1.374449,104.433526,120.043670,0.497664,15.640064,4162.918106,37542.461122,1.629246,9.813646,21.433241,4162.918106,21045.379569,336.446807,21.800901,0.070415,52111.490959,139.732492,0.466796,1.453451,132.273077,120.043670,104.433526,31919.279101,336.446807,15.640064,NaN,139.732492,0.0
min,0.000000,0.000000,-4274.979980,7.490000,0.000000,2.000000,1.000000,603.000000,2.000000,-33.937553,-158.025986,1.000000,1.000000,19.000000,0.000000,0.000000,1.000000,9.990000,-2.750000,1.000000,9.990000,7.490000,-4274.979980,1040.000000,19.000000,2.000000,NaN,9.990000,0.0
25%,2.000000,2.000000,7.000000,104.379997,0.000000,18.000000,3258.500000,725.000000,4.000000,18.265432,-98.446312,3258.500000,18057.000000,403.000000,5.400000,0.040000,45130.500000,50.000000,0.080000,1.000000,119.980003,104.379997,7.000000,23464.000000,403.000000,18.000000,NaN,50.000000,0.0
50%,3.000000,4.000000,31.520000,163.990005,1.000000,29.000000,6457.000000,19380.000000,5.000000,33.144863,-76.847908,6457.000000,36140.000000,627.000000,14.000000,0.100000,90260.000000,59.990002,0.270000,1.000000,199.919998,163.990005,31.520000,59405.000000,627.000000,29.000000,NaN,59.990002,0.0
75%,5.000000,4.000000,64.800003,247.399994,1.000000,45.000000,9779.000000,78207.000000,7.000000,39.279617,-66.370583,9779.000000,54144.000000,1004.000000,29.990000,0.160000,135389.500000,199.990005,0.360000,3.000000,299.950012,247.399994,64.800003,90008.000000,1004.000000,45.000000,NaN,199.990005,0.0
max,6.000000,4.000000,911.799988,1939.989990,1.000000,76.000000,20757.000000,99205.000000,12.000000,48.781933,115.263077,20757.000000,77204.000000,1363.000000,500.000000,0.250000,180519.000000,1999.989990,0.500000,5.000000,1999.989990,1939.989990,911.799988,99301.000000,1363.000000,76.000000,NaN,1999.989990,0.0


**Statistical Summary:** The `df.describe()` method generates descriptive statistics that summarize the central tendency, dispersion, and shape of the dataset’s distribution, excluding NaN values. It provides metrics such as mean, standard deviation, minimum, maximum, and percentiles for numerical columns.

### 2.5. Data dictionary

In [ ]:
# ---- Data Dictionary table -----------------------------------------------
# Role codes:
#   TARGET  = what we predict  (must NOT be used as a model input feature)
#   FEATURE = input the model is allowed to use at prediction time
#   ID      = identifiers, not predictive but useful for joins
#   PII     = personal data, must be dropped immediately
#   DROP    = too sparse or irrelevant to be useful

data_dict_rows = [
    # (Column Name, Meaning, Raw dtype, Role, Notes)

    # -- Order identifiers ---------------------------------------------------
    ('Order Id',                    'Unique order identifier',                      'int',      'ID',      'Primary key'),
    ('Order Item Id',               'Unique order-line identifier',                 'int',      'ID',      'One order can have multiple items'),
    ('Order Customer Id',           'Links order back to the customer',             'int',      'ID',      'Redundant with Customer Id; drop'),
    ('Order Item Cardprod Id',      'Product code via RFID reader',                'int',      'ID',      'Redundant with Product Card Id; drop'),
    ('Product Card Id',             'Product code',                                 'int',      'ID',      ''),
    ('Product Category Id',         'Product category code',                        'int',      'ID',      'Duplicate of Category Id; drop one'),
    ('Category Id',                 'Product category code',                        'int',      'ID',      ''),
    ('Department Id',               'Store department code',                        'int',      'ID',      ''),
    ('Customer Id',                 'Unique customer identifier',                   'int',      'ID',      ''),

    # -- Targets -------------------------------------------------------------
    ('Late_delivery_risk',          '1 = late, 0 = on time  [CLASSIFIER TARGET]',  'int',      'TARGET',  'Binary label derived from actual vs scheduled days'),
    ('Delivery Status',             'Human-readable delivery outcome',              'str',      'TARGET',  'Direct source of Late_delivery_risk - exclude from features'),
    ('Days for shipping (real)',    'Actual number of shipping days taken',         'int',      'TARGET',  'Used to compute delay_gap regression target - exclude'),

    # -- Delivery / temporal features ----------------------------------------
    ('Days for shipment (scheduled)', 'Promised delivery window in days',           'int',      'FEATURE', 'Known at order time - strong predictor'),
    ('order date (DateOrders)',     'Date the order was placed',                   'str->date','FEATURE', 'Parse to datetime; extract month and weekday'),
    ('shipping date (DateOrders)', 'Date and time shipment was dispatched',        'str->date','FEATURE', 'Only known post-dispatch - leakage risk for live scoring'),
    ('Shipping Mode',               'Standard / First / Second / Same Day class',  'str',      'FEATURE', 'Likely the strongest single categorical predictor'),

    # -- Financial features --------------------------------------------------
    ('Benefit per order',           'Net earnings on the order line (USD)',         'float',    'FEATURE', 'Used in Risk Score profit_weight calculation'),
    ('Order Profit Per Order',      'Net profit on the order (USD)',                'float',    'DROP',    'Exact duplicate of Benefit per order - drop one'),
    ('Sales per customer',          'Total sales value for the customer order',     'float',    'FEATURE', ''),
    ('Sales',                       'Sales value for this order line',              'float',    'FEATURE', 'Correlated with several other financial columns'),
    ('Order Item Total',            'Total amount for the order line',              'float',    'FEATURE', 'Correlated with Sales'),
    ('Order Item Discount',         'Absolute discount applied (USD)',              'float',    'FEATURE', ''),
    ('Order Item Discount Rate',    'Discount rate 0.0-0.25 (observed)',           'float',    'FEATURE', 'Appears as discrete bands; consider encoding'),
    ('Order Item Product Price',    'Unit price before discount',                   'float',    'FEATURE', ''),
    ('Order Item Profit Ratio',     'Profit / price ratio',                         'float',    'FEATURE', 'Some values < -1: extreme losses; flag as outlier indicator'),
    ('Order Item Quantity',         'Number of units ordered',                      'int',      'FEATURE', 'Range 1-5'),
    ('Product Price',               'Listed product price',                         'float',    'FEATURE', 'Correlated with Order Item Product Price'),

    # -- Categorical / geographic features -----------------------------------
    ('Type',                        'Payment type: DEBIT / TRANSFER / PAYMENT / CASH', 'str', 'FEATURE', ''),
    ('Customer Segment',            'Consumer / Corporate / Home Office',           'str',      'FEATURE', ''),
    ('Market',                      'Global delivery market: LATAM / Europe / etc.','str',      'FEATURE', ''),
    ('Order Region',                'Sub-region of delivery destination',           'str',      'FEATURE', 'More granular than Market'),
    ('Order Country',               'Destination country',                          'str',      'FEATURE', 'High cardinality; may need grouping'),
    ('Order City',                  'Destination city',                             'str',      'FEATURE', 'Very high cardinality - likely drop at modelling stage'),
    ('Order State',                 'Destination state or region',                  'str',      'FEATURE', ''),
    ('Department Name',             'Name of the fulfilling department',            'str',      'FEATURE', ''),
    ('Category Name',               'Product category name',                        'str',      'FEATURE', ''),
    ('Product Name',                'Name of the ordered product',                  'str',      'FEATURE', 'High cardinality; use Category Name instead'),
    ('Product Status',              '0 = in stock, 1 = out of stock',              'int',      'FEATURE', ''),
    ('Order Status',                'COMPLETE / PENDING / CANCELED / etc.',         'str',      'FEATURE', 'May contain post-delivery info; use with caution'),
    ('Latitude',                    'Latitude of the fulfilling store',             'float',    'FEATURE', ''),
    ('Longitude',                   'Longitude of the fulfilling store',            'float',    'FEATURE', ''),

    # -- PII (must be removed) -----------------------------------------------
    ('Customer Fname',              'Customer first name',                          'str',      'PII',     'No predictive value; drop immediately'),
    ('Customer Lname',              'Customer last name (8 missing)',              'str',      'PII',     'Drop'),
    ('Customer Email',              'Customer email address',                       'str',      'PII',     'Drop'),
    ('Customer Password',           'Masked password hash',                         'str',      'PII',     'Drop'),
    ('Customer Street',             'Customer street address',                      'str',      'PII',     'Drop'),
    ('Customer City',               'City of customer address',                     'str',      'PII',     'Drop - use Order City instead'),
    ('Customer Country',            'Country of customer',                          'str',      'PII',     'Drop - use Order Country instead'),
    ('Customer State',              'State of customer',                            'str',      'PII',     'Drop - use Order State instead'),
    ('Customer Zipcode',            'Customer zip code (3 missing)',               'float',    'PII',     'Drop'),

    # -- Drop (not useful) ---------------------------------------------------
    ('Order Zipcode',               'Destination zip code',                         'float',    'DROP',    '86.2% missing - too sparse to impute reliably'),
    ('Product Description',         'Free-text product description',               'float',    'DROP',    '100% missing - entirely empty column'),
    ('Product Image',               'URL to product image',                         'str',      'DROP',    'URL string with no predictive value'),
]

dd = pd.DataFrame(data_dict_rows, columns=['Column', 'Meaning', 'Raw Type', 'Role', 'Notes'])

for role in ['TARGET', 'FEATURE', 'ID', 'PII', 'DROP']:
    sub = dd[dd['Role'] == role]
    print(f'\n{"="*65}')
    print(f'  {role}  ({len(sub)} columns)')
    print(f'{"="*65}')
    for _, row in sub.iterrows():
        print(f'  {row["Column"]:<42} {row["Raw Type"]:<12} {row["Notes"]}')

print(f'\nTotal columns classified: {len(dd)}')

Before writing a single line of cleaning code, the first task is to understand what each column actually represents. Working from the official column descriptions provided with the dataset, the table below assigns every field a role label - Target, Feature, ID, PII, or Drop - which drives every decision in Step 2.

The two most important role distinctions are:

- **TARGET:** the label we want to predict - these columns must be excluded from input features to prevent data leakage.
- **PII:** Personally Identifiable Information (names, emails, passwords) - these must be removed before any analysis or sharing.

## 3. Data Cleaning & Formatting <a id="cleaning"></a>

Each cleaning decision follows the same three-question discipline:

1. **What is wrong?** (name the problem specifically).
2. **How widespread is it?** (a number, not just 'some').
3. **What is the safest fix?** (and why not something else).

Every step is documented so this notebook tells the full portfolio story.

### 3.1. Drop columns flagged as PII, DROP, or exact duplicates

In [ ]:
DROP_COLS = [
    # 100% and 86% missing columns - cannot be meaningfully imputed
    'Product Description',
    'Order Zipcode',

    # PII - personally identifiable information
    'Customer Fname', 'Customer Lname', 'Customer Email',
    'Customer Password', 'Customer Street',
    'Customer City', 'Customer Country', 'Customer State', 'Customer Zipcode',

    # URL with no predictive value
    'Product Image',

    # Exact column duplicates (confirmed r=1.0 in correlation check)
    'Order Profit Per Order',   # identical to Benefit per order
    'Product Category Id',      # identical to Category Id
    'Order Item Cardprod Id',   # identical to Product Card Id
    'Order Customer Id',        # redundant link to Customer Id
]

df = df_raw.drop(columns=DROP_COLS)
print(f'Columns remaining: {df.shape[1]}  (dropped {len(DROP_COLS)} columns)')

Data cleaning complete. Current shape: (180519, 49)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 49 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   Type                           180519 non-null  object        
 1   Days for shipping (real)       180519 non-null  int64         
 2   Days for shipment (scheduled)  180519 non-null  int64         
 3   Benefit per order              180519 non-null  float64       
 4   Sales per customer             180519 non-null  float64       
 5   Delivery Status                180519 non-null  object        
 6   Late_delivery_risk             180519 non-null  int64         
 7   Category Id                    180519 non-null  int64         
 8   Category Name                  180519 non-null  object        
 9   Customer City                  180519 non-null  object        
 10  Customer Country

### 3.2. Fix data types: parse date columns

In [ ]:
# The two date columns arrived as plain strings (dtype=object).
# Converting them to datetime64 unlocks .dt accessors for feature extraction
# and allows us to compute time-based features (month, weekday, lead time).

df['order date (DateOrders)']    = pd.to_datetime(df['order date (DateOrders)'],    errors='coerce')
df['shipping date (DateOrders)'] = pd.to_datetime(df['shipping date (DateOrders)'], errors='coerce')

print('Date column types after conversion:')
print(df[['order date (DateOrders)', 'shipping date (DateOrders)']].dtypes)
print(f'\nOrder date range: {df["order date (DateOrders)"].min().date()} '
      f'to {df["order date (DateOrders)"].max().date()}')
print(f'Null dates after parsing: {df["order date (DateOrders)"].isnull().sum()}')

### 3.3. Missing value audit

In [ ]:
missing = df.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)
mv = pd.DataFrame({'Missing Count': missing, 'Missing %': missing_pct})
remaining_missing = mv[mv['Missing Count'] > 0]

if len(remaining_missing) == 0:
    print('No missing values remain after dropping sparse columns.')
else:
    print('Columns with remaining missing values:')
    print(remaining_missing.to_string())

# DECISION: After dropping the 100% and 86% missing columns, no column
# has meaningful missingness. The 8 missing Customer Lname values were
# already removed with the PII drop. No imputation is required.
print(f'\nTotal NaN values in working dataset: {df.isnull().sum().sum()}')

### 3.4. Duplicate check

In [ ]:
full_dupes = df.duplicated().sum()
order_line_dupes = df.duplicated(subset=['Order Id', 'Order Item Id']).sum()

print(f'Fully identical rows         : {full_dupes:,}')
print(f'Duplicate (Order + Item) rows: {order_line_dupes:,}')

# DECISION: Zero duplicates on both checks. No action required.
# The dataset is already clean at the row level.

### 3.5. Check for impossible or suspect values

In [ ]:
print('Impossible value checks:')
print(f'  Days shipping (real) < 0 ..........: {(df["Days for shipping (real)"] < 0).sum():,}')
print(f'  Days scheduled < 0 ................: {(df["Days for shipment (scheduled)"] < 0).sum():,}')
print(f'  Order Item Quantity < 1 ...........: {(df["Order Item Quantity"] < 1).sum():,}')
print(f'  Order Item Discount < 0 ...........: {(df["Order Item Discount"] < 0).sum():,}')
print(f'  Profit Ratio < -1 (extreme loss) ..: {(df["Order Item Profit Ratio"] < -1).sum():,}')
print(f'  Profit Ratio > 1 (impossible gain) : {(df["Order Item Profit Ratio"] > 1).sum():,}')
print(f'  Sales < 0 ..........................: {(df["Sales"] < 0).sum():,}')

print()
print('DECISION - Profit Ratio < -1 (6,301 rows, ~3.5% of data):')
print('  These orders recorded a loss larger than the product price itself.')
print('  This is unusual but can reflect heavy discounting, returns, or write-offs.')
print('  SAFE FIX: create a binary indicator flag to preserve the signal rather than')
print('  silently deleting rows. The model can learn from extreme loss events.')

In [ ]:
# Create the outlier indicator flag.
# This is preferable to deletion because the model may learn that extreme-loss orders
# behave differently from normal orders.
df['profit_ratio_extreme_loss'] = (df['Order Item Profit Ratio'] < -1).astype(int)

print(f'Flag created. Rows flagged as extreme loss: '
      f'{df["profit_ratio_extreme_loss"].sum():,} '
      f'({df["profit_ratio_extreme_loss"].mean()*100:.1f}%)')

### 3.6. Engineer the regression target: `delay_gap`

In [ ]:
# delay_gap is the core regression target for Stage 3 of the pipeline.
# Formula: actual days taken minus scheduled days promised.
#   Positive value = arrived late (by that many days)
#   Zero           = exactly on schedule
#   Negative       = arrived early (advance shipping)

df['delay_gap'] = df['Days for shipping (real)'] - df['Days for shipment (scheduled)']

print('delay_gap distribution (days late or early):')
print(df['delay_gap'].value_counts().sort_index().rename('Count').to_frame().to_string())
print(f'\nMean delay  : {df["delay_gap"].mean():.2f} days')
print(f'Pct late (>0): {(df["delay_gap"] > 0).mean()*100:.1f}%')
print(f'Pct early (<0): {(df["delay_gap"] < 0).mean()*100:.1f}%')

### 3.7. Extract date features

In [ ]:
# Extract calendar features from the order placement date.
# These are safe to use in the model because they are known at the moment of ordering.

df['order_year']    = df['order date (DateOrders)'].dt.year
df['order_month']   = df['order date (DateOrders)'].dt.month
df['order_weekday'] = df['order date (DateOrders)'].dt.dayofweek  # 0=Monday, 6=Sunday

# The shipping date is available post-dispatch only, so we extract it purely for
# EDA visibility below but will NOT include it as a model feature.
# (Including it would leak future information into the prediction.)

print('Year range:', df['order_year'].min(), 'to', df['order_year'].max())
print('Orders per year:')
print(df['order_year'].value_counts().sort_index().to_string())

### 3.8. Final cleaned dataset summary

In [ ]:
print('=' * 50)
print('CLEANED DATASET SUMMARY')
print('=' * 50)
print(f'Rows         : {len(df):,}')
print(f'Columns      : {df.shape[1]}')
print(f'Missing vals : {df.isnull().sum().sum()}')
print()
print('Target variables:')
print(f'  Classifier : Late_delivery_risk (0/1)')
vc = df['Late_delivery_risk'].value_counts()
for v, cnt in vc.items():
    label = 'Late' if v == 1 else 'On time'
    print(f'    {label} ({v}) : {cnt:,} ({cnt/len(df)*100:.1f}%)')
print(f'  Regressor  : delay_gap (range: {df["delay_gap"].min()} to {df["delay_gap"].max()} days)')
print()
print('New engineered columns added:')
for col in ['delay_gap', 'profit_ratio_extreme_loss', 'order_year', 'order_month', 'order_weekday']:
    print(f'  {col}')

df.to_csv('dataco_cleaned.csv', index=False)
print('\nSaved to dataco_cleaned.csv')

## 4. Exploratory Data Analysis (EDA) <a id="eda"></a>
EDA is where raw numbers become a story. Rather than producing every possible chart, the goal is to answer four focused questions:

1. What does each key variable look like? (distributions)
2. How do variables relate to each other? (correlations)
3. What patterns connect to late delivery? (target analysis)
4. What might cause modelling problems? (class imbalance, collinearity, outliers)